cw-relayer/tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,175 @@
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type (
+	RateMsg struct {
+		Ref Rate `json:"get_ref"`
+	}
+
+	Rate struct {
+		Symbol string `json:"symbol"`
+	}
+
+	RateData struct {
+		Rate        string `json:"rate"`
+		ResolveTime string `json:"resolve_time"`
+		RequestId   string `json:"request_id"`
+	}
+
+	RefMsg struct {
+		RefData Symbol `json:"get_reference_data"`
+	}
+
+	Symbol struct {
+		SymbolPair [2]string `json:"symbol_pair"`
+	}
+
+	RefData struct {
+		Rate             string `json:"rate"`
+		LastUpdatedBase  string `json:"last_updated_base"`
+		LastUpdatedQuote string `json:"last_updated_quote"`
+	}
+
+	RefMsgBulk struct {
+		RefData Symbols `json:"get_reference_data_bulk"`
+	}
+
+	Symbols struct {
+		SymbolPairs [][2]string `json:"symbol_pairs"`
+	}
+)
+
+func (s *IntegrationTestSuite) TestQueryFromContract() {
+	grpcConn, err := grpc.Dial(
+		s.orchestrator.QueryRpc,
+		// the Cosmos SDK doesn't support any transport security mechanism
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+
+	s.Require().NoError(err)
+	defer grpcConn.Close()
+
+	testCases := []struct {
+		tc      string
+		prepare func() ([]byte, error)
+		assert  func(queryData []byte) bool
+	}{
+		{
+			tc: "query rate from contract",
+			prepare: func() ([]byte, error) {
+				msg := RateMsg{Ref: Rate{Symbol: "TEST"}}
+				data, err := json.Marshal(msg)
+				if err != nil {
+					return nil, err
+				}
+
+				return data, err
+			},
+			assert: func(queryData []byte) bool {
+				var resp RateData
+				err = json.Unmarshal(queryData, &resp)
+				if err != nil {
+					return false
+				}
+
+				s.Require().Equal(resp.Rate, "1234567890")
+
+				return true
+			},
+		},
+		{
+			tc: "query reference data in USD from contract",
+			prepare: func() ([]byte, error) {
+				msg := RefMsg{Symbol{[2]string{"TEST", "USD"}}}
+				data, err := json.Marshal(msg)
+				if err != nil {
+					return nil, err
+				}
+
+				return data, nil
```

Comment:
It looks like most of the "prepare" and "assert" functions are relatively the same. Are you unable to make a generalized function, and then make the "prepared" and "expected" values table-driven?
If we can, we should - more readability